### PR TITLE
expression: fix a bug that DML using caseWhen may cause schema change

### LINKED ISSUE #19387: `select sum(case when 1 then a end) from t group by a` causes schema change
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` sql
mysql> create table t4(a decimal(16, 2));
OK, 0 row affected

mysql> select sum(case when 1 then a end) from t4 group by a;
OK, 0 row in set (0.00 sec)

mysql> show create table t4;
+-------+------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                     |
+-------+------------------------------------------------------------------------------------------------------------------+
| t4    | CREATE TABLE `t4` (
  `a` decimal(65,2) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

create index idx on t4(a);
insert into t4 values (12.34);

... restart TiDB server ...

mysql> select * from t4 use index(idx) where a=12.34;
Empty set (0.00 sec)

mysql> select * from t4 use index() where a=12.34;
+-------+
| a     |
+-------+
| 12.34 |
+-------+
1 row in set (0.00 sec)

```

This bug can be fixed by https://github.com/pingcap/tidb/pull/19083.

### 2. What did you expect to see? (Required)

`decimal(16,2)` in `show create table`, that is, the schema won't change after executing the SQL.

### 3. What did you see instead (Required)

`decimal(16,2)` in `show create table`, that is, the schema changes after executing the SQL.
Because of that, the index is wrong because it does not match the schema.

### 4. What is your TiDB version? (Required)

v3.0.16, master

## SIG slack channel
[#sig-exec](https://slack.tidb.io/invite?team=tidb-community&channel=sig-exec&ref=high-performance)

## Score
- 300

## Mentor
 * @SunRunAway

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #19387 

Problem Summary:

1. `BuildCastFunction(expr.GetCtx(), foldedExpr, foldedExpr.GetType())` in `constant_fold.go` in wrong and needless.
`foldedExpr.GetType()` should be `expr.GetType()`.
2. In WrapCastForAggArgs, we change the pointer to fieldType. It's very dangerous.

### What is changed and how it works?

1. Remove `BuildCastFunction(expr.GetCtx(), foldedExpr, foldedExpr.GetType())`.
2. Use a safe way to change FieldType.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- fix a bug that DML using caseWhen may cause schema change

### CODE REVIEW

File: expression/aggregation/base_func.go [hunk 1]
```diff
@@ -357,17 +357,14 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
...
-		if col, ok := a.Args[i].(*expression.Column); ok {
-			col.RetType = types.NewFieldType(col.RetType.Tp)
+		// internally.
+		switch x := a.Args[i].(type) {
+		case *expression.Column:
+			x.RetType = a.RetTp
+		case *expression.ScalarFunction:
+			x.RetType = a.RetTp
+		case *expression.CorrelatedColumn:
+			x.RetType = a.RetTp
```
Comment (qw4990) [inline expression/aggregation/base_func.go:367]: Do we need to copy the `RetTp` for safety?

Comment (wjhuang2016) [reply expression/aggregation/base_func.go:367]: No, we can't 100% ensure it's not updated in other places. 
There are too many usages like here. If we use clone() every time, the performance would be worst.

**Review by qw4990 — COMMENTED** [review]
LGTM

**Comment by qw4990** [comment]
PTAL @wshwsh12

**Review by bb7133 — APPROVED** [review]
LGTM

**Comment by wshwsh12** [comment]
Please fix the ci.

**Comment by wshwsh12** [comment]
```

[2020-09-23T06:42:56.643Z] FAIL: integration_test.go:153: testIntegrationSuite.TestAggregation

[2020-09-23T06:42:56.643Z] 

[2020-09-23T06:42:56.643Z] integration_test.go:176:

[2020-09-23T06:42:56.643Z]     tk.MustQuery(sql).Check(testkit.Rows(output[i].Result...))

[2020-09-23T06:42:56.643Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:62:

[2020-09-23T06:42:56.643Z]     res.c.Assert(resBuff.String(), check.Equals, needBuff.String(), res.comment)

[2020-09-23T06:42:56.643Z] ... obtained string = "[110]\n"

[2020-09-23T06:42:56.643Z] ... expected string = "[110.0]\n"

[2020-09-23T06:42:56.643Z] ... sql:select sum(case when a <= 0 or a > 1000 then 0.0 else b end) from t, args:[]
```

**Comment by wjhuang2016** [comment]
Waiting for https://github.com/pingcap/tidb/pull/20017

**Pull request merged**